index.bs

```diff
@@ -346,10 +346,47 @@ interface mixin DatagramTransport {
 
 ## Procedures ## {#datagrams-transport-procedures}
 
-The <dfn>readDatagrams</dfn> algorithm is given a |transport| as parameter. It
-is defined by running the following steps:
-1. TODO
-1. [=ReadableStream/enqueue=] |data| in |transport|.`[[IncomingDatagrams]]`.
+To <dfn>pullDatagrams</dfn>, given a {{WebTransport}} object |transport|, run these steps:
+1. Let |queue| be |transport|'s [=[[IncomingDatagramsQueue]]=].
+1. If |queue| is empty, then:
+  1. Assert: |transport|'s [=[[IncomingDatagramsPullPromise]]=] is null.
+  1. Set |transport|'s [=[[IncomingDatagramsPullPromise]]=] to a new promise.
+  1. Return |transport|'s [=[[IncomingDatagramsPullPromise]]=].
+1. Let |bytes| and |timestamp| be |queue|'s first element.
+1. Remove the first element from |queue|.
+1. Let |chunk| be a new {{Uint8Array}} object representing |bytes|.
+1. [=ReadableStream/Enqueue=] |chunk| to |transport|'s [=[[IncomingDatagrams]]=].
+1. Return [=a promise resolved with=] undefined.
+
+To <dfn>receiveDatagrams</dfn>, given a {{WebTransport}} object |transport|, run these steps:
+1. Let |timestamp| be a timestamp representing now.
+1. Let |queue| be |transport|'s [=[[IncomingDatagramsQueue]]=].
+1. Let |duration| be |transport|'s [=[[IncomingDatagramsExpirationDuration]]=].
+1. If |duration| is null, then set |duration| to an [=implementation-defined=] value.
+1. Let |connection| be |transport|'s [=[[Connection]]=].
+1. While there are available incoming datagrams on |connection|: [[!WEB-TRANSPORT-HTTP3]]
+
+  Issue: We should probably use "session" instead of "connection", to support pooling.
+
+  1. Let |datagram| be the first datagram, and remove it from |connection|.
+  1. Let |timestamp| be a timestamp representing now.
+  1. Let |chunk| be a pair of |datagram| and |timestamp|.
+  1. Enqueue |chunk| to |queue|.
+1. Let |toBeRemoved| be the length of |queue| minus |transport|'s
+  [=[[IncomingDatagramsHighWaterMark]]=].
+1. If |toBeRemoved| is positive, remove first |toBeRemoved| elements from |queue|.
+1. While |queue| is not empty:
+  1. Let |bytes| and |timestamp| be |queue|'s first element.
+  1. If more than |duration| seconds have passed since |timestamp|, then remove the first element
+     from |queue|.
+  1. Otherwise, break this loop.
+1. If |queue| is not empty and |transport|'s [=[[IncomingDatagramsPullPromise]]=] is non-null, then:
+   1. [=resolve=] |transport|'s [=[[IncomingDatagramsPullPromise]]=] with undefined.
+   1. Set |transport|'s [=[[IncomingDatagramsPullPromise]]=] to null.
```

Comment:
I think we need to enqueue to `[[IncomingDatagrams]]` here. Otherwise I think the following scenario won't work:

1. WebTransport object created
2. Connection completes asynchronously sometime later
3. `read()` is called, causing...
4. `pullDatagrams` to be called
5. `queue` is empty, so `[[IncomingDatagramsPullPromise]]` is created and returned.
6. A datagram arrives, causing...
7. `receiveDatagrams` algorithm to run
8. `queue` becomes non-empty, so `[[IncomingDatagramsPullPromise]]` is resolved.
9. Step 7 of [ReadableStreamDefaultControllerCallPullIfNeeded](https://streams.spec.whatwg.org/#readable-stream-default-controller-call-pull-if-needed) executes.
10. No new read has been performed and nothing has been enqueued, so `pullDatagrams` is not called again.
11. The promise returned by `read()` never resolves.

Reply:
Nice catch, fixed.